### PR TITLE
service/state: Provide simple scaffold for `StateService` and implement `CoreAccess`

### PR DESCRIPTION
**Pull request by renaynay**
This PR is the first implementation of `StateAccessor` over core: `CoreAccess`. It will provide a `light node` with the ability to be initialised with a connection to a celestia-core endpoint over which it can query for account balances and submit transactions.

[ADR](https://github.com/celestiaorg/celestia-node/blob/main/docs/adr/adr-004-state-interaction.md) for reference.

Two notes: 
1. Although it was documented in the ADR, I decided *against* using the [lens](https://github.com/strangelove-ventures/lens) dependency as it is still in development  and introduced a bit more overhead / complexity than was necessary for two simple queries against the application.
2. It is extremely difficult to write unit tests that test the full functionality of `CoreAccess` as we lack a good way to mock the celestia-app for testing purposes. This will be discussed in the following week and will hopefully provide more clarity in how we can increase test coverage for this feature.

TODO: 
- [x] doc.go
- [ ] swamp test
- [x] start/stop hooks w/ DI
- [x] construction
- [x] lint
- [x] update ADR

### CODE REVIEW

File: node/services/config.go
Comment (adlerjohn) [inline node/services/config.go:25]: Can we document this now?

**Review by adlerjohn — COMMENTED** [review]
Lint error

File: node/services/config.go
Comment (adlerjohn) [reply node/services/config.go]: Documented in [b3b5707](https://github.com/celestiaorg/celestia-node/pull/420/commits/b3b5707466aa31385127857142f5ad8b834bcb1d)

**Review by liamsi — COMMENTED** [review]
> Although it was documented in the ADR, I decided against using the [lens](https://github.com/strangelove-ventures/lens) dependency as it is still in development and introduced a bit more overhead / complexity than was necessary for two simple queries against the application

Do you mind quickly updating the ADR accordingly in this PR?

File: node/state/core.go
Comment (liamsi) [inline node/state/core.go:21]: What does the comment refer to?

**Review by liamsi — APPROVED** [review]
👏🏼

File: node/state/core.go
Comment (renaynay) [reply node/state/core.go:21]: I'm passing nil here for `userInput` parameter, but not really sure about it.

File: service/state/core_access.go
Comment (vgonkivs) [inline service/state/core_access.go:47]: Dial will send error immediately If there even a small chance that server could start later than client. I'd suggest to add [WithBlock](https://pkg.go.dev/google.golang.org/grpc#WithBlock) option + context with Timeout. In this case you need [DialContext](https://pkg.go.dev/google.golang.org/grpc#DialContext) instead of Dial

**Review by adlerjohn — COMMENTED** [review]
Merge conflicts

**Comment by renaynay** [comment]
Rebased.

**Comment by renaynay** [comment]
FYI the test is currently broken as mock celestia-app isn't working.

**Review by adlerjohn — COMMENTED** [review]
Lint errors!

File: service/state/core_access.go
Comment (adlerjohn) [inline service/state/core_access.go:32]: If this never fails, is there a reason for it to return an `error`?

File: service/state/core_access.go
Comment (adlerjohn) [inline service/state/core_access.go:53]:
```suggestion
func (ca *CoreAccessor) Stop(context.Context) error {
```

File: service/state/core_access.go
Comment (adlerjohn) [inline service/state/core_access.go:65]: Can you add docstrings to public methods? For example, it's completely non-obvious that `Balance()` returns the balance of a particular address.

Comment (renaynay) [reply service/state/core_access.go:65]: they're documented in the interface

Comment (adlerjohn) [reply service/state/core_access.go:65]: Oh I see

**Comment by renaynay** [comment]
Test is failing due to failure in mock celestia-app

**Review by adlerjohn — COMMENTED** [review]
Is CI expected to fail, or no?

**Comment by renaynay** [comment]
Yes @adlerjohn

**Review by adlerjohn — APPROVED** [review]
utACK

File: node/services/config.go
Comment (Wondertan) [inline node/services/config.go:71]: I guess this is the name for a default keyring account name. Do we really need a way to configure its name? It does not seem to change anything meaningful for users.

To handle a default Node key we can:
* Have a name constant "default" which is the default for everyone and this is the key that is autogenerated on the first start of the node, so users would need to top it up with tokens to use it. This is the most simplest approach
* Have an API that allows users to generate/import a key/account and set it as default with an API. This is an approach we should get to eventually. 

In both cases, I don't see why users would want to change their default keyring name through a config.

**Review by Wondertan — CHANGES_REQUESTED** [review]
image

Mostly looks good and straightforward. Good that the implementation turned out to be quite small even without a lens. T

here are only a few things that require changes, other than that are mostly nits, questions, and discussions on how things should look further.

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go:66]: I am in favour of documenting them only in interface as there are less places for inconsistent docs if we do end up changing them in one place and not the other.

File: go.mod
Comment (renaynay) [reply go.mod:57]: Nothing we can do 😭  thanks, cosmos-sdk 😭

File: docs/adr/adr-004-state-interaction.md
Comment (renaynay) [reply docs/adr/adr-004-state-interaction.md]: resolved?

File: service/state/interface.go
Comment (renaynay) [reply service/state/interface.go:18]: could use `types.AccAddress` ?

```go
// AccAddress a wrapper around bytes meant to represent an account address.
// When marshaled to a string or JSON, it uses Bech32.
type AccAddress []byte
```

wdyt @Wondertan ?

File: node/components.go
Comment (renaynay) [reply node/components.go:41]: Yeah I can do that. The main issue is -- we wanted full node to be able to *provide* state rather than use a core accessor to access it. I think that having that functionality available in light node for now is good. I can also add it to *bridge* node since it already receives a core endpoint anyway. Wouldn't hurt to implement core_access there as well. Wdyt regarding *full* node?

File: go.sum
Comment (renaynay) [reply go.sum:65]: :/

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go]: resolved?

File: service/state/core_access.go
Comment (Wondertan) [reply service/state/core_access.go:62]: I agree with you, but we should also be on the same page with @liamsi

File: node/services/config.go
Comment (renaynay) [reply node/services/config.go]: Have taken the first approach. Resolved?

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go]: See https://github.com/celestiaorg/celestia-node/pull/420/commits/8959e791e5355100dca4221558943657da1bedfa

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go]: i like having req defined separately.

File: node/components.go
Comment (renaynay) [reply node/components.go:30]: Because it's the only way to tell right now whether light node is running with a core connection.

Comment (renaynay) [reply node/components.go:30]: This is to prevent adding an additional flag.

File: node/state/core.go
Comment (renaynay) [reply node/state/core.go:18]: can we expose the path as a method on `keystore.Keystore` then? I can't just pass a string to `lightComponents`, I'd have to pass the whole keystore bc of fx

File: service/state/core_access.go
Comment (liamsi) [reply service/state/core_access.go:68]: Hmm, can you have a look on how they look on godoc? I'd make the decision depending on that: how are the implemented methods displayed and is there a connection from the impl to the interface on godoc.

File: node/components.go
Comment (Wondertan) [reply node/components.go:31]: Ok, I see

File: node/state/core.go
Comment (liamsi) [reply node/state/core.go:18]: I think this is to get the user's password for the file: https://github.com/cosmos/cosmos-sdk/blob/20735e792c85a0d84ca32e97a0a3c814bcc3529f/crypto/keyring/keyring.go#L671

File: node/components.go
Comment (Wondertan) [reply node/components.go]: Until we don't implement the FullNode state computation there won't be a way for a full node operator to access the state, considering that we don't support running light and full nodes together(ref to a different ports discussion), so I think it is fine for a full node to rely on a Core as long as this is not a requirement.

File: service/state/core_access.go
Comment (Wondertan) [reply service/state/core_access.go:87]: Refer to sdk_tx pkg for a description of a difference between modes.  

Regarding sync and async methods: I would keep only one method with options that allow setting the mode(this is what I am doing with PublicAPI)

File: node/components.go
Comment (Wondertan) [reply node/components.go:31]: SGTM

File: service/state/core_access.go
Comment (Wondertan) [reply service/state/core_access.go:68]: Good point, cc @renaynay

File: node/state/core.go
Comment (liamsi) [reply node/state/core.go:18]: Can we manually test this before merging? I think in the SDK this is used with Stdin. The userInput is fetched from this context:
https://github.com/cosmos/cosmos-sdk/blob/fece3f0485dd418a2bc54ecc61f17da3303192c7/simapp/simd/cmd/root.go#L46

File: service/state/core_access.go
Comment (liamsi) [reply service/state/core_access.go:68]: ref: https://github.com/celestiaorg/celestia-node/issues/426#issue-1131238570

File: node/state/core.go
Comment (Wondertan) [reply node/state/core.go:18]: Yes. I think it is fine to pass the Keystore to fx(I guess we are doing this already) and to state.Accessor contructor

File: service/state/interface.go
Comment (Wondertan) [reply service/state/interface.go:18]: I think it is fine. I also checked the SDK and they have an interface that AccAddress supports and I think we should use the interface instead.

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go]: Resolved?

File: node/state/core.go
Comment (renaynay) [reply node/state/core.go:18]: I manually tested it and it does not panic.

Comment (renaynay) [reply node/state/core.go:18]: what is the userInput used for?

Comment (liamsi) [reply node/state/core.go:18]: To get the user's input for the password, which is used to encrypt the file. If it does not panic, it's OK for now. We should probably support all the other backends as well though.

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go]: I changed default to `MODE_BLOCK`

should I add this method as well

```go
func (ca *CoreAccessor) SubmitTxWithOptions(ctx context.Context, tx Tx, mode sdk_tx.BroadcastMode) (*TxResponse, error) {
	txResp, err := apptypes.BroadcastTx(ctx, ca.coreConn, mode, tx)
	if err != nil {
		return nil, err
	}
	return txResp.TxResponse, nil
}
```

Comment (renaynay) [reply service/state/core_access.go]: Pls check @Wondertan: https://github.com/celestiaorg/celestia-node/pull/420/commits/f05d22b8eba7eede502ec4dc270e636db1b8be94

File: node/state/core.go
Comment (renaynay) [reply node/state/core.go]: Resolved https://github.com/celestiaorg/celestia-node/pull/420/commits/59d2a7c6091607692aad8506047024ab6b615164 ?

File: node/components.go
Comment (renaynay) [reply node/components.go]: Resolved https://github.com/celestiaorg/celestia-node/pull/420/commits/46f49d2feec485267a3f995907c9240ef628ef70 ?

File: service/state/interface.go
Comment (renaynay) [reply service/state/interface.go]: Resolved https://github.com/celestiaorg/celestia-node/pull/420/commits/46f49d2feec485267a3f995907c9240ef628ef70 ?

Comment (Wondertan) [reply service/state/interface.go]: No, I meant [this](https://github.com/celestiaorg/cosmos-sdk/blob/master/types/address.go#L104) interface

File: node/components.go
Comment (Wondertan) [reply node/components.go]: No. What about Bridge? Don't see a reason for it to be restricted to state access, so let's put this in baseComponents.
> Any node should be able to access the state

File: service/state/core_access.go
Comment (Wondertan) [reply service/state/core_access.go]: Not necessary part pf this PR.
> should I add this method as well

**Comment by Wondertan** [comment]
@renaynay, this PR is fully broken

File: node/state/core.go
Comment (Wondertan) [inline node/state/core.go:19]:
```suggestion
	// TODO @renaynay: Include option for setting custom `userInput` parameter with
```

File: service/state/interface.go
Comment (Wondertan) [reply service/state/interface.go]: Also, pls make a local alias to that interface, like you did for other types.

File: node/components.go
Comment (renaynay) [reply node/components.go]: https://github.com/celestiaorg/celestia-node/pull/420/commits/9e3e46bc16cfd28172f5927e1a32856380a27eeb

Now?

**Review by Wondertan — APPROVED** [review]
Ok lets go

File: service/state/core_access.go
Comment (liamsi) [reply service/state/core_access.go:69]: What was the conclusion here?

File: service/state/core_access_test.go
Comment (liamsi) [inline service/state/core_access_test.go:6]: I'd prefer a skipped test instead of commented out. Also an issue that describes the necessary steps to unblock the test.

**Review by liamsi — APPROVED** [review]
LFG

**Review by adlerjohn — APPROVED** [review]
Issues for the open conversations should be opened, but not going to block on them.

File: service/state/core_access_test.go
Comment (renaynay) [reply service/state/core_access_test.go:6]: Deleted test, tracking [here](https://github.com/celestiaorg/celestia-node/issues/570)

File: service/state/core_access.go
Comment (renaynay) [reply service/state/core_access.go:69]: Conclusion: if it implements an interface, those methods do not need to be documented in the implementation (only in the interface) as it reduces surface-area for having discrepancies in documentation. Any methods outside of interface impl on that struct must be documented, though.

**Pull request merged**
